### PR TITLE
sending libraries user is following down in a separate list

### PR DESCRIPTION
**Pull request by 2is10**
and no longer mixing `"access"` into each `LibraryInfo`
@stkem

### CODE REVIEW

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala [hunk 1]
```diff
@@ -146,34 +146,43 @@ class LibraryController @Inject() (
...
-      val (owner, numKeeps) = db.readOnlyMaster { implicit s =>
-        (basicUserRepo.load(library.ownerId), keepRepo.getCountByLibrary(library.id.get))
-      }
-      val info = LibraryInfo.fromLibraryAndOwner(library, owner, numKeeps, None)
-      val memInfo = if (mem.lastViewed.nonEmpty) Json.obj("access" -> mem.access, "lastViewed" -> mem.lastViewed) else Json.obj("access" -> mem.access)
-      Json.toJson(info).as[JsObject] ++ memInfo
+    val (libsWithMemberships, libsWithAllInvites) = libraryCommander.getLibrariesByUser(request.userId)
+    val libsWithInvites = for ((lib, invites) <- libsWithAllInvites.groupBy(_._2).mapValues(_.map(_._1))) yield {
+      (invites.sorted.last, lib) // only show one invite per library - the one with highest access
     }
```
Comment (2is10) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala:152]: These four lines are equivalent to 149-155 above—just renaming and trying to improve readability.

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala [hunk 2]
```diff
@@ -146,34 +146,43 @@ class LibraryController @Inject() (
...
-    val invitesToShow = librariesWithInvites.groupBy(x => x._2).map { lib =>
-      val invites = lib._2.unzip._1
-      val highestInvite = invites.sorted.last
-      (highestInvite, lib._1)
-    }.toSeq
-
-    val libsFollowing = for ((mem, library) <- librariesWithMemberships) yield {
-      val (owner, numKeeps) = db.readOnlyMaster { implicit s =>
-        (basicUserRepo.load(library.ownerId), keepRepo.getCountByLibrary(library.id.get))
-      }
```
Comment (2is10) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala:160]: This old code was opening a database session per library (owned, followed, and invited). New code below uses only one database session and uses `basicUserRepo.loadAll` to make a bulk memcache call.

**Comment by 2is10** [comment]
I’ll probably wait a couple of hours, at least, to merge this, to avoid breaking existing open long-lived browser sessions. In reality, that’s probably just Danny, but still worthwhile.

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala (hunk 2 above)
Comment (stkem) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala:160]: :+1:

**Comment by stkem** [comment]
:-) lgtm
Makes me wonder if we should have some kind of force reload build into kifi.com if your session is older than some amount of time?

**Comment by 2is10** [comment]
That’s a good idea. Or a way to signal old apps to reload just when we know we’re making a backwards incompatible change.

File: kifi-backend/modules/shoebox/test/com/keepit/controllers/website/LibraryControllerTest.scala [hunk 3]
```diff
@@ -657,8 +658,7 @@ class LibraryControllerTest extends Specification with ShoeboxTestInjector {
                     |"numKeeps":0,
                     |"numFollowers":0,
                     |"kind":"user_created",
-                    |"inviter":{"id":"${user2.externalId}","firstName":"Baron","lastName":"B","pictureName":"0.jpg","username":"bhsu"},
-                    |"access":"read_insert"
+                    |"inviter":{"id":"${user2.externalId}","firstName":"Baron","lastName":"B","pictureName":"0.jpg","username":"bhsu"}
```
Comment (2is10) [inline kifi-backend/modules/shoebox/test/com/keepit/controllers/website/LibraryControllerTest.scala:661]: @andrewconner And this even had test coverage! Unfortunately, the inviter was also another library’s owner, which gave me a false sense of security.

**Pull request merged**